zio-http/src/test/scala/zhttp/http/HttpDataSpec.scala

```diff
@@ -41,7 +61,78 @@ object HttpDataSpec extends DefaultRunnableSpec {
               assertM(res)(equalTo("abc\nfoo"))
             },
           ),
+          suite("UnsafeAsync")(
+            testM("sending last msg should finish the processing with 1 msg.") {
+              val unsafeChannel = new UnsafeReadableChannel {
+                override def read(): Unit = println("called read.")
+              }
+
+              val probe  = HttpData.UnsafeAsync { cb =>
+                val producer = cb(unsafeChannel)
+
+                producer(new UnsafeContent(new DefaultLastHttpContent(Unpooled.EMPTY_BUFFER)))
+              }
+              val result = probe.toByteBufStream(ByteBufConfig(4)).runCount
+              assertM(result)(equalTo(1L))
+            },
+            testM("Sending a chunk without last msg the stream will not complete") {
+              val unsafeChannel = new UnsafeReadableChannel {
+                override def read(): Unit = ()
+              }
+
+              val probe  = HttpData.UnsafeAsync { cb =>
+                val producer = cb(unsafeChannel)
+                producer(
+                  new UnsafeContent(new DefaultHttpContent(Unpooled.copiedBuffer("abc".toArray.map(_.toByte)))),
+                )
+              }
+              val result = probe.toByteBufStream(ByteBufConfig(4)).runCount
+              assertM(result)(equalTo(1L))
+            } @@ timeout(1 seconds) @@ failing,
+            testM("sending 2 msg should finish the processing with same number of messages.") {
+
+              val p             = Promise[Unit]()
+              val unsafeChannel = TestUnsafeReadableChannel(p)
+
+              val probe  = HttpData.UnsafeAsync { cb =>
+                val producer = cb(unsafeChannel)
+                producer(
+                  new UnsafeContent(new DefaultHttpContent(Unpooled.copiedBuffer("abc".toArray.map(_.toByte)))),
+                )
+                p.future.onComplete(_ => producer(new UnsafeContent(new DefaultLastHttpContent(Unpooled.EMPTY_BUFFER))))
+
+              }
+              val result = probe.toByteBufStream(ByteBufConfig(4)).runCount
+              assertM(result)(equalTo(2L))
+            },
+            testM("sending multiple msg should finish the processing with same number of messages.") {
+
+              val p1            = Promise[Unit]()
+              val p2            = Promise[Unit]()
+              val queue         = Queue(p1, p2)
+              val unsafeChannel = TestUnsafeReadableChannel2(queue)
+
+              val probe  = HttpData.UnsafeAsync { cb =>
+                val producer = cb(unsafeChannel)
+                producer(
+                  new UnsafeContent(new DefaultHttpContent(Unpooled.copiedBuffer("abc".toArray.map(_.toByte)))),
```

Comment:
could we please simplify this test too much code.

Reply:
re-implemented using generators